### PR TITLE
Rework Texture Interface

### PR DESCRIPTION
**Pull request by ebiederstadt**
Changes the texture interface use std::variant<> instead of individual load functions. This has the following benefits:
- Reduces code duplication. No need to specify GL_TEXTURE_2D (for example) in the constructor and the load function.
- Clearer user API. Users don't need to worry about which load function to call, they just pass in their data and that's it.
- Easy upgrade path. In the future we can add different texture types without adding another load function.

I have tested these changes in fork of bg-helloworld and they seem to be working nicely. If you like this direction I can have a PR ready for givderender quickly :)

### CODE REVIEW

**Review by lakinwecker — CHANGES_REQUESTED** [review]
Marking this as "request changes" until we have time to discuss it (hopefully today).

My primary points are:
1. I don't think image loading should be part of this library. OpenGL doesn't care how you get your image, and I don't think poggles should either.
2. While this interface is unequivocally better than what we had before, it doesn't solve my primary issue with all of this, which is that we should not have to change this file to be able to change the way that poggles deals with textures. Ideally, poggles doesn't care _where_ the texture comes from. 
3. This all will be made moot by the changes I would like to propose that we make to givderender.

None of that prevents us from merging it, I'm happy to do so because of the fact that this improves the situation. I just want to make sure we do it with the above in mind.

File: source/poggles/texture.cpp [hunk 1]
```diff
@@ -1,94 +1,125 @@
...
+            if (textureFromData.channels == 1) {
+              format = GL_RED;
+            } else if (textureFromData.channels == 2) {
+              format = GL_RG;
+            } else if (textureFromData.channels == 3) {
+              format = GL_RGB;
+            } else if (textureFromData.channels == 4) {
+              format = GL_RGBA;
+            }
+            GLenum target = textureFromData.target.value_or(m_original_target);
```
Comment (JeremyAdamHart) [inline source/poggles/texture.cpp]: m_originalTarget doesn't get updated in the texture loading, and there's no check to make sure that m_originalTarget is the same as GL_TEXTURE_2D. This is a problem in the original code, and definitely not something you introduced, but it's something we should probably fix :grimacing: 

```suggestion       
            m_originalTarget = textureFromData.target.value_or(m_original_target);

            glTexParameteri(m_originalTarget, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
            glTexParameteri(m_originalTarget, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
            glTexParameteri(m_originalTarget, GL_TEXTURE_WRAP_S, GL_REPEAT);
            glTexParameteri(m_originalTarget, GL_TEXTURE_WRAP_T, GL_REPEAT);

            GLenum format = GL_RGB;
            if (textureFromData.channels == 1) {
              format = GL_RED;
            } else if (textureFromData.channels == 2) {
              format = GL_RG;
            } else if (textureFromData.channels == 3) {
              format = GL_RGB;
            } else if (textureFromData.channels == 4) {
              format = GL_RGBA;
```

**Review by JeremyAdamHart — CHANGES_REQUESTED** [review]
I think this is a fine way to do things, but I wonder, what are the advantages to doing this instead of having a function that creates (or updates) a texture?

There are many ways to load and create textures (For instance tex parameters can be set differently than how we currently have them set), and I'd have a preference of supporting those options outside of the details of the texture class. Instead of having these functions as a part of the texture class, we could have something like this:

`
loadTextureFromFile(poggles::texture& tex, std::filesystem::path filename){
    // Insert details here
}
`
Part of why I prefer this paradigm is that it can be done in any library, and the only thing that needs to change here is having the target made public. This adds some potential issues if people modify it randomly though which I don't love, but there may be a good way of handling that too

**Comment by ebiederstadt** [comment]
There is no particular advantage with a function that creates a texture, I just went with this approach because it follows the class design that the rest of the library follows. I would be happy if we created functions to load textures instead of making users use the constructor (especially because it sounds like that is the direction that we will be taking this library anyways).

File: source/poggles/texture.cpp (hunk 1 above)
Comment (ebiederstadt) [reply source/poggles/texture.cpp]: This is out out date now :)

**Comment by ebiederstadt** [comment]
Jeremy made all the changes we were talking about to this PR, so it should be good to go now :)

**Review by JeremyAdamHart — APPROVED** [review]
Yup! Let's merge it in!
Just some notes on the merge, this is a breaking change, so make sure you include feat!: at the beginning of the squashed commit message.
Additionally, since we updated the version.txt file, CI won't automatically do the version bump on its own, so you'll need to manually bump it to 0.3.4. Once you get that commit in it's all good to merge

**Comment by benjaminulmer** [comment]
I went ahead and fixed the version so we can merge whenever now

File: include/poggles/texture.h [hunk 2]
```diff
@@ -12,36 +15,30 @@ namespace poggles
...
+  GLenum target;
 
 private:
-  GLenum m_original_target;
   texture_handle m_texture_handle;
 };
-}  // namespace poggles
+
+auto loadFromFile(texture& tex,
+                  std::filesystem::path const& filepath,
```
Comment (lakinwecker) [inline include/poggles/texture.h]: I mildly prefer to use `poggles::texture::fromFile` as the name for these, but its a minor preference

**Pull request merged**